lib/rex/proto/http/proxy.rb

```diff
@@ -0,0 +1,492 @@
+    self.request_timeout        = 30
+    # If the keep-alive session is pushing to the client, monitor and forward the traffic
+    # self.monitor_thread         = Rex::ThreadFactory.spawn("HTTP Proxy Monitor Thread", true) { monitor_clients }
+  end
+
+  #
+  # Returns the hardcore alias for the HTTP proxy service
+  #
+  def self.hardcore_alias(*args)
+    "#{(args[0] || '')}#{(args[1] || '')}"
+  end
+
+  #
+  # HTTP server.
+  #
+  def alias
+    super || "HTTP Proxy"
+  end
+
+  #
+  # Listens on the defined port and host and starts monitoring for clients.
+  #
+  def start
+
+    self.listener = Rex::Socket::TcpServer.create(
+      'LocalHost' => self.listen_host,
+      'LocalPort' => self.listen_port,
+      'Context'   => self.context,
+      'SSL'				=> self.ssl,
+      'SSLCert'		=> self.ssl_cert,
+      'Comm'      => self.comm
+    )
+
+    # Register callbacks
+    self.listener.on_client_connect_proc = Proc.new { |cli|
+      on_client_connect(cli)
+    }
+    self.listener.on_client_data_proc = Proc.new { |cli|
+      on_client_data(cli)
+    }
+
+    self.listener.start
+  end
+
+  #
+  # Terminates the monitor thread and turns off the listener.
+  #
+  def stop
+    self.clients.map {|cli| cli.close_session}
+    self.listener.stop
+    self.listener.close
+  end
+
+
+  #
+  # Waits for the HTTP service to terminate
+  #
+  def wait
+    self.listener.wait if self.listener
+  end
+
+  #
+  # Closes the supplied client, if valid.
+  #
+  def close_client(cli)
+    cli.close_session
+    listener.close_client(cli)
+    self.clients.select! {|x| x != cli}
+  end
+
+  ##
+  # Convenience methods for headers
+  ##
+
+  #
+  # Adds Server headers if they dont exist
+  #
+  def add_headers(resp,headers)
+    headers.each do |k,v|
+      resp[k] = v unless resp[k]
+    end
+  end
+
+  #
+  # Replaces server headers only if exist
+  #
+  def sub_headers(resp,headers)
+    headers.each do |k,v|
+      resp[k] = v if resp[k]
+    end
+  end
+
+  #
+  # Puts in all headers, whether existing or not
+  #
+  def force_headers(resp,headers)
+    add_response_headers(resp,headers)
+    sub_response_headers(resp,headers)
+  end
+
+  #
+  # Sends a 404 error to the client for a given request.
+  #
+  def send_e404(cli, request)
+    resp = Response::E404.new
+
+    resp['Content-Type'] = 'text/html'
+
+    resp.body =
+      "<html><head>" +
+      "<title>404 Not Found</title>" +
+      "</head><body>" +
+      "<h1>Not found</h1>" +
+      "The requested URL #{html_escape(request.resource)} was not found on this server.<p><hr>" +
+      "</body></html>"
+
+    # Send the response to the client like what
+    cli.send_response(resp)
+  end
+
+  # Rewrite proxy elements of request to target request
+  # Determine whether or not to use SSL for outbound
+  def proxy_header_rewrite(request)
+    if request.uri.match(/^http/i)
+
+      # Determine if we need to use SSL
+      request.headers['SSL'] = (request.uri[4].downcase == 's')
```

Comment:
`request.uri[4].downcase == 's'` isn't intuitive.

Consider `request.uri.start_with?('https')`

Reply:
Predicting a future ticket that makes me very sad, I'd throw a down case before the start_with

Reply:
Agreed.

Reply:
Roger, wilco